lib/semian/adapter.rb

```diff
@@ -0,0 +1,34 @@
+module Semian
+  module Adapter
+    def semian_identifier
+      raise NotImplementedError.new("Semian adapters must implement a `semian_identifier` method")
+    end
+
+    private
+
+    def acquire_semian_resource(scope:)
+      return yield if @resource_acquired # Prevent a single adapter from acquiring the resource twice by nesting calls
```

Comment:
because comments are often considered a code smell and easily get out of sync with the code. but if you think that it really improves the code here, then ok... I would prefer to have a helper method with a more descriptive name rather than a comment.. something like
```ruby
return yield if same_adapter_already_acquired_resource?
```